:book: :sparkles: :people_holding_hands: add proposal for Node Bootstrapping working group (#11407)
kubernetes-sigs/cluster-api: docs/community/20241112-node-bootstrapping.md

  - https://github.com/kubernetes-sigs/cluster-api/issues/9157
  - https://youtu.be/0AhA4Box3MM?si=IfKJfMWmlA9EW7ri&t=172
---

# Node Bootstrapping Working Group ("CAPI-NoBo")

This document briefly outlines the scope, communication media, and
stakeholders for a Working Group dedicated to the Provisioning aspect of
Node Bootstrapping.

Provisioning, in this context, refers to configuring and customising the
node operating system to prepare it to serve as a ClusterAPI cluster node.

## User Story and Problem Statement

As a Linux Distribution maintainer aiming to support ClusterAPI, or to improve
my existing support of ClusterAPI in my distribution, I would like to be able to
cleanly integrate my contributions into the node bootstrapping process without
interfering with other implementations.
I would like my contributions to be as re-usable as possible across different
bootstrap providers.
I would like to read and to follow documentation, guidelines, and specifications
on the above.
I would like to be able to do this to deliver choice, variety, and innovation to
users of my own distribution as well as to the larger ClusterAPI ecosystem, and

> to enable other distribution vendors to do the same.

Author: That's a good point, I made this part more specific.

**Problem statement / Example issues**

Currently, the kubeadm bootstrap provider implements two provisioning systems:
cloud-init and ignition.

> As there is currently no abstraction of node configuration, both implementations

Reviewer: What do we mean by "node configuration"? Is it different than "provisioning"?
I so, maybe we should clarify what this term means since "configuration" is a very overloaded term.
If not, I suggest we use consistent terminology to avoid further confusion around this matter.

Author: We could agree on "OS provisioning" and could define it as "Configuration of the operating system at provisioning time".

must cover generic configuration as well as specific config language concerns.
This leads to duplicate code and to maintenance overhead.
Also, other providers like e.g. MicroK8s must implement their own OS configuration
generators since kubeadm's node provisioning implementations cannot be re-used.
MicroK8s currently only supports cloud-init.

Outside of ClusterAPI, both [cloud-init](https://github.com/canonical/cloud-init)

> and [ignition](https://github.com/coreos/ignition) provisioning is widely adopted

Reviewer: ```suggestion
and [ignition](https://github.com/coreos/ignition) provisioning are widely adopted
```

across distributions.
While cloud-init is used by general-purpose Linux distributions like Ubuntu/Debian,
SUSE Linux, Alma, Rocky, Fedora, and Red Hat Enterprise Linux, ignition is popular
with special-purpose distributions like Fedora CoreOS / Red Hat CoreOS, SuSE MicroOS
/ SLE Micro / Kalpa, and Flatcar Container Linux.

> It is likely that more provisioning systems exist; a clean separation and guidelines
> will make it easier to add support to ClusterAPI as well as to share implementations
> across bootstrap providers.

Reviewer: ```suggestion
It is likely that more provisioning systems exist; a clean separation between bootstrap and provisioning
and guidelines on how each is developed will make it easier to add support to ClusterAPI as well as to
share implementations across bootstrap providers.
```

> **Compatibility and API guarantees**

Reviewer: Does this mean this work group can't work on any changes which may result in breaking changes? IMO this will severely limit the possible solutions we could come up with.

Maybe we should mention that backward-compatible changes are _preferable_ and that in case breaking changes are introduced, a well-documented migration path with a proper deprecation process should be put in place?

Author: That's a good point, I amended the section accordingly.

ClusterAPI has been stable for multiple years and is in widespread production use.
Proposals and implementations of this working group must ensure existing integrations
continue to work.
Any changes in the kubeadm bootstrap provider in particular must uphold guarantees and
expectations set by the current implementation, and must continue to support current use
cases.

## Scope

> 1. Propose architectural improvements that ease integration of node configuration

Reviewer: Again, not clear on what "node configuration" refers to.

Author: I mixed two concerns into one (well, two) words. I've amended 1. accordingly.

     implementations.
   Decouple generic bootstrapping from configuration language specific concerns.

> 2. Deliver an example implementation in the kubeadm bootstrapper.

Reviewer: ```suggestion
2. Deliver an example implementation in the kubeadm bootstrap provider.
```

Better to use the [official terminology](https://cluster-api.sigs.k8s.io/reference/glossary#bootstrap-provider) to avoid confusion. Ditto all occurrences below.

3. Approach other bootstrappers and provide help and guidance for separating provisioning
   and re-using provisioning implementations.

## Communication

We will join the [regular ClusterAPI meetings](https://github.com/flatcar-hub/cluster-api?tab=readme-ov-file#-community-discussion-contribution-and-support)
and share updates on our work in the "Feature Groups" section, which is a regular part of

> the ClusterAPI meetings)

Author: Will add a link to our [WG minutes doc](https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit?tab=t.0#heading=h.ac1b03gmo5ph) and update the markdown file as soon as we aligned on a meeting cadence as aligned on in the [Nov 27 Capi Office Hours](https://docs.google.com/document/d/1GgFbaYs-H6J5HSQ6a7n4aKpk0nDLE2hgG2NSOM9YIRw/edit?tab=t.0#heading=h.faoavlb0ym66).

Will resolve this conversation to make the PR more reviewable.

with special-purpose distributions like Fedora CoreOS / Red Hat CoreOS, SuSE MicroOS
/ SLE Micro / Kalpa, and Flatcar Container Linux.
It is likely that more provisioning systems exist; a clean separation between bootstrap and provisioning
and guidelines on how each is developed will make it easier to add support to ClusterAPI as well as to
share implementations across bootstrap providers.

**Compatibility and API guarantees**

ClusterAPI has been stable for multiple years and is in widespread production use.
Proposals and implementations of this working group must ensure to not excert operational
risk on existing integrations.
Any changes in the kubeadm bootstrap provider in particular must uphold guarantees and
expectations set by the current implementation and must continue to support current use
cases OR must provide means for current use cases to continue functioning for an appropriate
migration period, as well as define a clear path for migration.

## Scope

1. Propose architectural improvements that abstract system configuration with the goal of
   reducing duplicate functionality in current implementations, easing development of new configuration

>    features, and simplifying integration of new configuration systems for node bootstrapping.

Reviewer 2: I'm also interested in this idea in general (primarily adding a comment to this spot to have a thread). We're currently using metal3 to do bare metal provisioning, and have started work on a new bare metal oriented provider, both of which have a similar problem that most others do not: deploying outside of a cloud platform that take over most of the basic configuration.

To apply custom configuration it's still most convenient to use configuration systems like cloud-init etc., as that makes it a lot easier for users to provide custom images. This leads to two problems that are related to the scope of this WG:
- While systems like cloud-init allow you to provide "vendor" configuration in addition to the user configuration, the options for the former are somewhat limited (don't know about Ignition). In our experience it was necessary to use the bootstrap provider in order to apply configuration files, e.g. for advanced networking setups. This breaks the separation between infrastructure and Kubernetes configuration, and gets even more complex and tangled up when mixing in variables provided on a per-machine basis (possible with metal3 for example).
- Bare metal providers (or providers for platforms that do not come with enough automation) need to implement explicit support for configuration systems in order to use them to apply their configuration. If the system doesn't support providing multiple configurations (e.g. the inconsistent vendor/user config in cloud-init), the infra provider needs to merge configuration, making it even more complicated.

Both of these could be improved by allowing to extend the abstract configuration before turning it into a system-specific format.

I'm not sure if this topic came up already (I haven't read all the comments here or on the google doc), but I think this could be a good opportunity to make some improvements in CAPI geared towards on-prem use.

cc @kashifest - in case you're interested from a metal3 perspective

Author: This sounds great - and I second the sentiment of making this a focus item in the working group. I've added a "Maintainability, Backwards Compatibility, and Extensibility" track to the WG agenda: https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit?usp=sharing and took the liberty of adding you as a point of contact for these concerns.

Will close this thread to ease review of this PR as aligned on in the [Nov 27 office hours](https://docs.google.com/document/d/1GgFbaYs-H6J5HSQ6a7n4aKpk0nDLE2hgG2NSOM9YIRw/edit?tab=t.0#heading=h.faoavlb0ym66).

Author: (Oh and @schrej  I've added you to the WG proposal as a Stakeholder, please let me know if that works for you)

Reviewer 3: I totally agree with @schrej's point of view here

stakeholders for a Working Group dedicated to the Provisioning aspect of
Node Bootstrapping.

Provisioning, in this context, refers to configuring and customising the
node operating system to prepare it to serve as a ClusterAPI cluster node.

## User Story and Problem Statement

As a Linux Distribution maintainer aiming to support ClusterAPI, or to improve
my existing support of ClusterAPI in my distribution, I would like to be able to
cleanly integrate my contributions into the node bootstrapping process without
interfering with other implementations.
I would like my contributions to be as re-usable as possible across different
bootstrap providers.
I would like to read and to follow documentation, guidelines, and specifications
on the above.
I would like to offer a choice of node bootstrapping configuration systems to users,
enable distribution vendors to extend the choces available by adding new
configuration systems, and ease maintenance and enable extension of currently
supported systems for the ClusterAPI developer community.

**Problem statement / Example issues**

Currently, the kubeadm bootstrap provider implements two provisioning systems:
cloud-init and ignition.

> As there is currently no abstract way to pass configuration to  node bootstrapping,

Author: That's a good point, we might as well work out a more precise glossary in the WG. It will come in handy down the road. I added a "glossary" track to our WG agenda doc: https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit?tab=t.0#heading=h.ed54x9s57jp1

> both configuration system implementations - ignition and cloud-init - must cover

Author: I'll use "OS provisioning" to be more precise.

generic configuration as well as config language specific concerns.
This leads to duplicate code and to maintenance overhead.

> Also, other providers like e.g. MicroK8s must implement their own OS configuration

Author: Same - I'll use "OS provisioning". Thanks for the pointer!

> generators since kubeadm's node bootstrapping implementations cannot be re-used.

Author: OS provisioning indeed. Good catch!

across distributions.
While cloud-init is used by general-purpose Linux distributions like Ubuntu/Debian,
SUSE Linux, Alma, Rocky, Fedora, and Red Hat Enterprise Linux, ignition is popular
with special-purpose distributions like Fedora CoreOS / Red Hat CoreOS, SuSE MicroOS
/ SLE Micro / Kalpa, and Flatcar Container Linux.
It is likely that more provisioning systems exist; a clean separation between bootstrap and provisioning
and guidelines on how each is developed will make it easier to add support to ClusterAPI as well as to
share implementations across bootstrap providers.

**Compatibility and API guarantees**

ClusterAPI has been stable for multiple years and is in widespread production use.
Proposals and implementations of this working group must ensure to not excert operational
risk on existing integrations.
Any changes in the kubeadm bootstrap provider in particular must uphold guarantees and
expectations set by the current implementation and must continue to support current use
cases OR must provide means for current use cases to continue functioning for an appropriate
migration period, as well as define a clear path for migration.

## Scope

> 1. Propose architectural improvements that abstract system configuration with the goal of
>    reducing duplicate functionality in current implementations, easing development of new configuration
>    features, and simplifying integration of new configuration systems for node bootstrapping.

Reviewer 2: How about wording this a bit more explicitly, to avoid some of the overloaded terms:
```suggestion
1. Propose architectural improvements to the Machine provisioning process that separate Kubernetes bootstrapping tools (kubeadm, microk8s) from generating OS-distribution-specific initialization configuration (cloud-init, ignition) to avoid duplicate implementations, ease development of new configuration features and simplify implementing new providers for both kinds of tools.
```

>    features, and simplifying integration of new configuration systems for node bootstrapping.

Reviewer: @schrej I like this except "initialization configuration" because we already call the same thing "provisioning" elsewhere and I think we should try to have one term for each concern since this area of the project is confusing as it is.

>    features, and simplifying integration of new configuration systems for node bootstrapping.

Reviewer 2: The term is also used to describe the entire process of bringing up a new Machine, including creating the infrastructure, configuring the OS and setting it up as a K8s node. That's why I tried avoiding it here.

>    features, and simplifying integration of new configuration systems for node bootstrapping.

Author: It's not just (necessarily) OS / distribution specific initialisation - e.g. with Ignition one could perform significant customisation like setting the root partition filesystem, creating / adding partitions, using encrypted FS / encrypted volumes, and the like. Ignition runs in the initrd so supports significant customisation options.

I'd like to stick with "OS provisioning" for the time being, will update the section accordingly.

This document briefly outlines the scope, communication media, and
stakeholders for a Working Group dedicated to the Provisioning aspect of
Node Bootstrapping.

Provisioning, in this context, refers to configuring and customising the
node operating system to prepare it to serve as a ClusterAPI cluster node.

## User Story and Problem Statement

As a Linux Distribution maintainer aiming to support ClusterAPI, or to improve
my existing support of ClusterAPI in my distribution, I would like to be able to
cleanly integrate my contributions into the node bootstrapping process without
interfering with other implementations.
I would like my contributions to be as re-usable as possible across different
bootstrap providers.
I would like to read and to follow documentation, guidelines, and specifications
on the above.
I would like to offer a choice of node bootstrapping configuration systems to users,
enable distribution vendors to extend the choices available by adding new
configuration systems, and ease maintenance and enable extension of currently
supported systems for the ClusterAPI developer community.

**Problem statement / Example issues**

> Currently, the kubeadm bootstrap provider implements two systems to configure a node's
> operating system at provisoining time (referred to as "OS provisioning" from here on):
> cloud-init and ignition.
> As there is currently no abstract way to pass OS provisioning configuration to node
> bootstrapping, both OS provisioning system implementations - ignition and cloud-init - must cover
> both generic node bootstrapping as well as OS provisioning system specific concerns.
> This leads to duplicate code and to maintenance overhead.
> Also, other providers like e.g. MicroK8s must implement their own OS provisioning
> systems since kubeadm's implementations is specific to kubeadm and cannot be re-used.
> MicroK8s currently only supports cloud-init.

Reviewer: ```suggestion
As there is currently no OS provisioning configuration abstraction, the kubeadm bootstrap provider
is tightly coupled with cloud-init and Ignition. Furthermore, the Ignition implementation is built
on top of the cloud-init one. This makes it hard to develop provisioning implementations
independently and makes it hard to reuse code effectively between bootstrap implementations: Other
bootstrap providers such as MicroK8s must implement their own OS provisioning, and MicroK8s
currently only supports cloud-init.

In addition, the current design mixes bootstrap-related provisioning code (such as the commands
necessary for executing the kubeadm binary on a host) with generic OS customizations (such as disk
partitioning or Linux user configuration).
```

I suggest this rephrasing since IMO the current phrasing doesn't emphasize the extent of the current issues.

Our main problems:

- Bootstrap is tightly coupled with Provisioning. This is the biggest issue IMO.
- The Ignition implementation is built on top of the cloud-init implementation. This means that we not only have a tight coupling between bootstrap and provisioning, **in addition** Ignition fully depends on cloud-init.

The implications of these problems:

- We can't modify cloud-init or Ignition code without affecting kubeadm-related code or vice versa.
- We can't modify cloud-init code without affecting Ignition code.
- We can't flexibly combine bootstrap providers with provisioning implementations (regardless of the question of whether provisioning should be embedded inside the bootstrap provider or be fully independent).
- Provisioning config is "abused" by users who want to perform generic OS customization (unrelated to CAPI bootstrap, i.e. disk partitioning) and have no well-defined mechanism for doing so.

I've explained all of the above in detail under the [Motivation section](https://docs.google.com/document/d/1Fz5vWwhWA-d25_QDqep0LWF6ae0DnTqd5-8k8N0vDDM/edit?tab=t.0#heading=h.wcf1gbhz0yw1) of the WIP design proposal (I think it's relevant to read that section even if we don't end up adopting the specific proposed design).

Author: This is great! Thank you for elaborating.

/ SLE Micro / Kalpa, and Flatcar Container Linux.
It is likely that more provisioning systems exist; a clean separation between bootstrap and provisioning
and guidelines on how each is developed will make it easier to add support to ClusterAPI as well as to
share implementations across bootstrap providers.

**Compatibility and API guarantees**

ClusterAPI has been stable for multiple years and is in widespread production use.
Proposals and implementations of this working group must ensure to not exert operational
risk on existing integrations.
Any changes in the kubeadm bootstrap provider in particular must uphold guarantees and
expectations set by the current implementation and must continue to support current use
cases OR must provide means for current use cases to continue functioning for an appropriate
migration period, as well as define a clear path for migration.

## Scope

1. Propose architectural improvements that abstract OS provisioning configuration with the goal of
   reducing duplicate functionality in current implementations, easing development of new OS provisioning
   features, and simplifying integration of new configuration systems for OS provisioning.

> 2. Deliver an example implementation in the kubeadm bootstrap provider.

Reviewer: ```suggestion
2. Deliver an example implementation which works with the kubeadm bootstrap provider.
```

The current phrasing implies that provisioning is embedded inside bootstrap -- which is one problem we are trying to address so I'm not sure we want to prescribe this assumption in advance. We may well decide that we want to keep things as they are, but let's at least discuss things first.

Author: Good point; and I believe Fabrizio and you are in alignment with this.

---
title: ClusterAPI Node Bootstrapping Working Group
authors:
    - "@t-lo"
reviewers:
    - "@elmiko"
    - "@eljohnson92"
    - "@johananl"
    - "@tormath1"
    - "@fabriziopandini"

>     - "@sbueringer"

Reviewer 4: happy to be included as reviewer as well

creation-date: 2024-11-12
last-updated: 2024-11-15
status: proposed
see-also:
  - https://github.com/kubernetes-sigs/cluster-api/issues/5294
  - https://docs.google.com/document/d/1Fz5vWwhWA-d25_QDqep0LWF6ae0DnTqd5-8k8N0vDDM/edit
  - https://github.com/kubernetes-sigs/cluster-api/issues/9157
  - https://youtu.be/0AhA4Box3MM?si=IfKJfMWmlA9EW7ri&t=172
  - https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit
---

# Node Bootstrapping Working Group ("CAPI-NoBo")

This document briefly outlines the scope, communication media, and
stakeholders for a Working Group dedicated to the Provisioning aspect of
Node Bootstrapping.

> Provisioning, in this context, refers to configuring and customising the

Reviewer 5: Suggesting to change to American English just because later on is used. But the suggestion can be inverted if you prefer the British version.

```suggestion
Provisioning, in this context, refers to configuring and customizing the
```

> node operating system to prepare it to serve as a ClusterAPI cluster node.

Reviewer 5: ```suggestion
node's operating system to prepare it to serve as a ClusterAPI cluster node.
```

## User Story and Problem Statement

As a Linux Distribution maintainer aiming to support ClusterAPI, or to improve
my existing support of ClusterAPI in my distribution, I would like to be able to
cleanly integrate my contributions into the node bootstrapping process without
interfering with other implementations.
I would like my contributions to be as re-usable as possible across different
bootstrap providers.
I would like to read and to follow documentation, guidelines, and specifications
on the above.

> I would like to offer a choice of node bootstrapping configuration systems to users,

Reviewer 4: can we include an illustrative example? e.g "I would like to enable the bootstrap API consumer to express intent to fine tune nodes by setting vm.dirty_ratio while not having to reimplement the logic to setup kubelet and let the machine become a Node" if that's a good representative?

Author: This is a great idea!
Probably the not a perfect example though, as setting sysctls solely involves the OS layer - i.e. it does not strictly fulfil a provisioning need of the bootstrap provider. IOW we can provision the nodes perfectly fine and alternatively set the sysctl later, e.g via an operator. Also, this opens a whole can of works as we would later need to validate the intent (and let the provisioning fail if it doesn't work) etc. 

How about something more fundamentally provisioning related like disks set-up or disk partitioning? E.g. adding a required storage device / file system to a mount point? I'll add something in the doc.

enable distribution vendors to extend the choices available by adding new
configuration systems, and ease maintenance and enable extension of currently
supported systems for the ClusterAPI developer community.

**Problem statement / Example issues**

> As there is currently no OS provisioning configuration abstraction, the kubeadm bootstrap provider

Reviewer 4: it's probably worth mentioning as part of the context that our API also has a knob that enables full customization via custom userdata secret

risk on existing integrations.
Any changes in the kubeadm bootstrap provider in particular must uphold guarantees and
expectations set by the current implementation and must continue to support current use
cases OR must provide means for current use cases to continue functioning for an appropriate
migration period, as well as define a clear path for migration.

## Scope

1. Propose architectural improvements that abstract OS provisioning configuration with the goal of
   reducing duplicate functionality in current implementations, easing development of new OS provisioning
   features, and simplifying integration of new configuration systems for OS provisioning.
2. Deliver an example implementation which works with the kubeadm bootstrap provider.
3. Approach other bootstrap providers and provide help and guidance for separating provisioning
   and re-using provisioning implementations.

## Communication

* Meeting schedule: Fortnightly; every first and third Thursday of the month at 17:00 UTC
  (WG [calendar](https://calendar.google.com/calendar/u/0?cid=OTBkMjJjZGU0OTcyZjI0OGQ2NTE2YTk2ZGUwNWVmNjI1NTM2NDRmZmYyNjFlMjE1MGY1ZjIyOTU0NmQ1OWQ0MUBncm91cC5jYWxlbmRhci5nb29nbGUuY29t),

>       [iCal](https://calendar.google.com/calendar/ical/90d22cde4972f248d6516a96de05ef62553644fff261e2150f5f229546d59d41%40group.calendar.google.com/public/basic.ics)).

Reviewer 6: ```suggestion
      [iCal](https://calendar.google.com/calendar/ical/90d22cde4972f248d6516a96de05ef62553644fff261e2150f5f229546d59d41%40group.calendar.google.com/public/basic.ics).
```

---
title: ClusterAPI Node Bootstrapping Working Group
authors:
    - "@t-lo"
reviewers:
    - "@elmiko"
    - "@eljohnson92"
    - "@johananl"

>     - "@tormath1"

Reviewer 7: cc @vincepri @enxebre @sbueringer @chrischdi

>     - "@tormath1"

Author: Added you and Stefan. Will add others as they drop by in the PR comments and provide feedback.

  - https://youtu.be/0AhA4Box3MM?si=IfKJfMWmlA9EW7ri&t=172
---

# Node Bootstrapping Working Group ("CAPI-NoBo")

This document briefly outlines the scope, communication media, and
stakeholders for a Working Group dedicated to the Provisioning aspect of
Node Bootstrapping.

Provisioning, in this context, refers to configuring and customising the
node operating system to prepare it to serve as a ClusterAPI worker cluster node.

## User Story and Problem Statement

As a Linux Distribution maintainer aiming to support ClusterAPI, or to improve
support of ClusterAPI in my distribution, I would like to improve ClusterAPI's
node bootstrapping process and provide specifications on ClusterAPI's expectations
and requirements for provisioning-time configuration and customisation.

Initial focus of the working group is on separating node bootstrapping and node
OS configuration / customisation ("node provisioning" from here on).
At the time of writing, bootstrapping and provisioning are architecturally
intertwined; both are implemented in the bootstrap provider.
This increases maintenance effort and effectively prevents re-using provisioning

> features across bootstrap providers.

Reviewer 7: I think that this statement is somehow more hinting a solution (separation) than a problem.

Base on my understanding, we are trying to solve two problems:
- to offer to the users more combinations of provisioner/bootstrappers.
- to ensure those combinations can evolve freely, taking befits from the greatest an the latest the underlying technologies

But IMO, users don't care about separation at all. 
In fact, users will always need both a provisioner and a bootstrapper to get a working cluster.

If instead we split provisioner and a bootstrapper into two separate components, probably we are not making a good service both to the users, nor to ourself (as project maintainers), nor to the community that builds on top of Cluster API:

- we are adding to the system another dimension of complexity, that makes this change highly disruptive for users
- we are imaging a change that will impact not only the users, but the entire of the CAPI ecosystem, think e.g. to the CAPI operator, or all the products built on top of Cluster API.
- we are adding to the users an additional operational burden (they have to take care of yet another component/provider, its upgrades, its compatibility matrix, its CVEs etc etc). 
- we are imaging a change so complex and extensive that, being very pragmatic, the chances that this change might happen drastically drop (e.g. it will impact clusterctl, clusterclass, runtime extensions, all the controllers, almost everything; in fact this is why this problem is laying around for so long)

Frankly speaking, this seems too much to digest for everyone. 

The good news, it that I think that it is possible to look at the problem above from a different angle and find a better solution.

How to solve the two problem above (offer more combinations/ensure those combinations can evolve freely) without introducing separation?

That's up to the working group to figure out, but If I can give my two cents, what we are looking for is a solution that doesn't impact users, but makes it easier to implement and evolve different provisioners (e.g. we want to bump ignition!), while re-using existing bootstrappers (e.g. kubeadm).

I think that the key point in the sentence above is software re-use

So, a possible solution might be in a combination of federated development/separated repositories (providers as we call it in CAPI) + something that makes it easy to reuse code e.g. making kubeadm provider a library (details TBD of course, but I hope you get the idea)

Please also see https://docs.google.com/document/d/1Fz5vWwhWA-d25_QDqep0LWF6ae0DnTqd5-8k8N0vDDM/edit?disco=AAABX57uOnU where I make other considerations about the same point

Happy to chat about it if you are interested on this angle

Author: These are good points - will rework the proposal accordingly. Admittedly I wrote the first draft with the existing proposal in mind.

Thank you for your review!

Reviewer: Thanks @fabriziopandini. I completely understand the concerns you've raised and am likewise worried by them. I'm also 100% for maximizing code reuse.

If I understand correctly, what you are proposing is to keep provisioners (e.g. Ignition) tightly coupled with bootstrappers (e.g. kubeadm) and offer multiple bootstrap providers for each bootstrapper, one for each provisioner. So, for example, we would have a **kubeadm + cloud-config** bootstrap provider and a **kubeadm + Ignition** bootstrap provider in separate repositories.

If that's what you meant, I'm seeing the following problems with this approach:

1. The cartesian product of bootstrappers and provisioners can grow very large. True, with 1 bootstrapper and 2 provisioners we would have only 2 repositories to maintain, but what if we have 3 bootstrappers and 3 provisioners? That's **9 repositories** already. AFAICT this isn't premature optimization since we seem to already have [8 supported bootstrappers](https://cluster-api.sigs.k8s.io/reference/providers.html?highlight=k3s#control-plane) and at least 3 provisioners (cloud-init and Ignition, with CloudBase Init either implemented or planned - not sure). That's 24 repos if we don't separate.

2. Even if we decide the number of repos is manageable, it's not clear to me how we would achieve good code reuse when we have cloud-config fields embedded in types such as `KubeadmConfigSpec`. Example:

```yaml
apiVersion: controlplane.cluster.x-k8s.io/v1beta1
kind: KubeadmControlPlane
metadata:
  name: capi-azure-control-plane
  namespace: default
spec:
  kubeadmConfigSpec:
    ...
    diskSetup:
      filesystems:
      - device: /dev/disk/azure/scsi1/lun0
        extraOpts:
        - -E
        - lazy_itable_init=1,lazy_journal_init=1
        filesystem: ext4
        label: etcd_disk
      ...
      partitions:
      - device: /dev/disk/azure/scsi1/lun0
        layout: true
        overwrite: false
        tableType: gpt
```

In the snippet above, `diskSetup` is a field under `KubeadmConfigSpec`. This field is [specific to cloud-init](https://cloudinit.readthedocs.io/en/latest/reference/modules.html#disk-setup). That means `KubeadmConfigSpec` assumes cloud-init is the only provisioner. True, we could implement "translators" under the hood for Ignition and any other provisioner, but there is no guarantee for feature parity among all provisioners so IMO it's only a matter of time until we run into issues where some feature is supported by cloud-init but not by other provisioners.

I think we have a major design flaw in the current core API (mixing bootstrap and provisioning in the data model). True, it will be very disruptive to change this, but from my understanding you're concerned about building more complexity (e.g. Ignition 3) on top of a suboptimal foundation, and to me it seems we'd be doing exactly that if we keep the API as is and keep adding layers on top of it.

Also, I know that in the past some maintainers have expressed similar concerns: https://github.com/kubernetes-sigs/cluster-api/pull/4172#discussion_r600098540

To summarize, I think that if we avoid tackling the design problem head on and look for ways around it, we'd find ourselves in a similar situation later on, only then we'd be disrupting even more areas of the projects.

I'm open to changing my mind on this :slightly_smiling_face:

Reviewer 7: @johananl that's a complex discussion
you clearly went miles down this path, and everyone else have to catch up (this is what the working group is for).
what I'm asking now is just to be open to other options and to not tight this working group to your work, because this will prevent other's contributions.

Also, please note that, I'm bringing this up mostly because, based on my experience, this could become an issue for the working group down the line when broader consensus will be required.

WRT to the initial assumption of the use cases we have in front of us, I get that theoretically we have to address a possible high number of combinations, but from the other side, as a matter of fact only the ignition team has raised this concern so far.

I'm not sure why, but I think that we should consider that either not all the combinations of providers/bootstrappers makes sense or simply there is no demand for many of those combinations (but getting a better understanding of the problem scope is again part of the working group goal IMO)

WRT to the possible solution, frankly speaking I think it is too early to get into the weed of different options without a better understanding of the context; also this PR is probably not the right place for those discussions.

But just to avoid misunderstanding, I'm well aware of the fact that the current KubeadmConfig API now is not in an ideal state for what we are discussing (as a matter of fact it wasn't designed with re-use in mind). However, what I'm advocating for is to explore options for making re-use possible while minimising the number of breaking changes/impact for users.

But finding a solution for this problem is a discussion that I will be happy to have if the charter of this working group opens up to such options.

Reviewer: Hey @fabriziopandini. Thanks for your follow up.

Of course I'm open to discussing other solutions and would love to get contributions from others 🙂 I hope I didn't write anything which implies otherwise.

The specific solution in my WIP proposal definitely isn't the only one, and it also isn't a solution I had in mind in advance but rather what I could come up with after gathering a list of current issues and concerns from both users and maintainers around bootstrap and provisioning. I'm not attached to this proposal and am happy to drop it or replace it with a better proposal right away if we decide that's the way to go.

To be clear, the reason I became involved with this work initially was to improve the integration between Flatcar Container Linux and CAPI. So, with my Microsoft hat on I'd be very glad to avoid doing any major reworking of core CAPI pieces if it's not absolutely necessary. We'd be happy to simply keep Ignition supported, add support for Ignition v3 and work towards graduating Ignition support from experimental mode to a stable "GA".

However, I also have the OSS contributor hat, and with that hat on I always try to do what's right for the projects I'm involved in rather than solving ad-hoc issues which stand in my way while working on use cases relevant mainly to my company.

The reason I've gone into the weeds of #5294 and came up with this proposal is that previously we heard very clearly from some of the maintainers that before we can consider making Ignition a fully-supported, "GA" provisioner, we must tackle #5294. If that's no longer the case or if I somehow misunderstood the intention there -- great, as this would make our work around Ignition much shorter and easier 👍 

I'm also in favor of deferring the technical discussions to the work group.

I understand your point (as well as Stefan's) about keeping the API stable. That's very important indeed. The only reason I proposed changing core API objects is that as far as I understood the CAPI maintainers (or at least some of them) required that we cleanly separate bootstrap from provisioning, and I couldn't find a way to truly separate the two without touching the API because the two concerns are unfortunately entangled in the API.

Let's continue this discussion in the first iteration of the work group then 🙂

Reviewer 8: great discussion here, i tend to agree with Fabrizio and Stefan about the need to not disrupt the API and user experience too much. with that said though, i think it would be great if the working group can explore multiple options to see if there is a way we can use the current APIs/contracts, or perhaps add to them instead of modifying, but as noted above this is technical discussion we should have in the WG.

for this proposal, i think we just need to be clear about the goals of the WG and our desire to find solutions that do not require re-architecting core pieces of the project.

Author: I've captured our discussion and included deep links to here in the WG agenda doc: https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit?usp=sharing

Will close this thread to ease review of this PR as aligned on in the [Nov 27 office hours](https://docs.google.com/document/d/1GgFbaYs-H6J5HSQ6a7n4aKpk0nDLE2hgG2NSOM9YIRw/edit?tab=t.0#heading=h.faoavlb0ym66).
